rfcs/0000-multiple-kuadrant-instances.md

```diff
@@ -0,0 +1,152 @@
+
+Given a gateway, Kuadrant needs to find out which Kuadrant's instance is managing that specific gateway. By design, Kuadrant knows it is only one. There are at least two options to implement that mapping:
+* Read all Kuadrant CR objects and the first one that matches label selector.
+  * This approach works as long as the control plane ensures that each gateway is matched by only one kuadrant gateway selector. The control plane must reject any new kuadrant instance matching a gateway already "taken" by other kuadrant instance.
+* Add annotation in the gateway with a value of the Name/Namespace of the Kuadrant CR.
+  * This approach is commonly used. Requires annotation management.
+
+
+### Just one external auth stage and one rate limiting stage
+Kuadrant configures the gateway with a single external authorization stage (backed by Authorino) and a single external rate limiting stage (backed by Limitador).
+Multiple rate limit or authN/AuthZ stages involving multiple instances of Authorino and Limitador can be implemented, technically speaking.
+Until there is a real use case and it is strictly necessary, this scenario is discarded. The main reason is about complexity. It is already complex enough to reason about rate limiting and auth services having a single stage. Adding multiple rate limiting stages, or hitting multiple Limitador instances in a single stage (doable with the WASM module) makes it too complex to reason about observed behavior. Currently there is no use case to require that complex scenario.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Multitenancy is not a requested capability from users. Usually ingress gateways are shared resources managed by cluster administrators and a cluster may have only few of them. It is also a cluster admin task to route traffic to the ingress gateway. Cluster users usually do not control the life cycle of the ingress gateways in order to have their own Kuadrant instance.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+
+The gateway is the top class entity in the design, not the policy. The API protection happens at the gateway and the configuration needs to be done at the gateway. This kuadrant instance design protects the gateway isolating them from other instances (mis)configurations or traffic spikes.
+
+- What other designs have been considered and what is the rationale for not choosing them?
+
+TODO
+
+- What is the impact of not doing this?
+
+This design is a step forward in a consistent API to protect other service APIs. It makes easier to protect any API, no matter traffic nature. Either north-south or east-west. It makes easier to have the scenario where cluster users deploy their own (not ingress) gateways and enable API protection declaratively.
+
+# Prior art
+[prior-art]: #prior-art
+
+TODO
+
+# Unresolved questions
```

Comment:
It might already be work in progress, but a specification of how errors (and which error cases) would be surfaced to the user would be good. For instance, a policy is applied to a `HTTPRoute` alright. The route gets an additional `Gateway` wired to it… what's the behavior? How does the user know?

Reply:
> a policy is applied to a `HTTPRoute` alright. The route gets an additional `Gateway` wired to it… what's the behavior? How does the user know?

Currently, the addition of a new `Gateway` to a `HTTPRoute` targeted by a policy is not reflected in the policy status, just like any other event of modifying the `HTTPRoute` that is handled successfully by the controller. The controller basically only reports _"HTTPRoute is protected"_ or reconciliation errors. But even with the error messages, they are not always helpful to users.

I think this is already important today and even more so in a context of multiple Kuadrant instances per cluster. Users could benefit from the info of which Gateways have been configured for a policy, which were expected to but have not (and why not), in the policy status. Currently, users can only partially work that out by reading in the annotations of the Gateways which policies affect the Gateway.

Reply:
https://github.com/Kuadrant/kuadrant-operator/issues/140